enhancements/builds/multi-arch-builds.md

```diff
@@ -0,0 +1,456 @@
+flowchart
+A(Developer merges changes to SCM) --> B(SCM notifies OpenShift webhook)
+B --> C(OpenShift webhook triggers OpenShift BuildConfig)
+C --> E
+D(Developer triggers OpenShift BuildConfig) --> E
+R(Developer creates OpenShift Build) --> E
+
+E(openshift-controller-manager starts N+1 build pods) --> F(Build pod pushes image to registry)
+E --> N(Build pod pushes image to registry)
+N --> G
+E --> P(Build pod pushes image to registry)
+P --> G
+F --> G(Build pod pushes manifest list to registry)
+G --> Q(Build pod tags manifest list in registry)
+Q --> H(Build pod untags images in registry)
+
+Q --> I(Developer updates pod specs)
+I --> J(OpenShift restarts pods)
+
+Q --> K(Registry notifies OpenShift webhook) --> L(OpenShift triggers DeploymentConfig)
+L --> J
+
+Q --> M(Developer pulls images)
+```
+
+### API Extensions
+
+Build and BuildConfig objects (build.openshift.io/v1) will be extended to
+include a target platform list. If set, the build controller will attempt the
+build for the specified platforms. If the platform list is not set in the Build
+or BuildConfig, the build controller should check for a list in its
+OpenShiftControllerManagerConfig, encoded in the `config` configMap in the
+`openshift-controller-manager` namespace. If that is also unset, the build
+controller will attempt to build for all of the architectures it detects in the
+cluster. The build controller will update the Build object to always include
+the target platform list.
+
+Build and BuildConfig objects (build.openshift.io/v1) will be extended to also
+include a list of target platforms for which failures will not cause the build
+to fail. A failed platform-specific build on any subset of the listed platforms
+will not be considered a failure of the larger build attempt, but the manifest
+list produced by the final build pod will not include images for platforms
+whose builds failed.
+
+BuildStatusOutput objects will be extended to include information which would
+be needed in addition to the target platform when constructing an entry in a
+manifest list which refers to the image described by the BuildStatusOutput.
+
+Build object (build.openshift.io/v1) status fields will be extended to include
+a list, keyed by target platform string, to a structure containing per-target
+BuildStatus values.
+
+### Implementation Details/Notes/Constraints [optional]
+
+* Add a target platforms []string field to github.com/openshift/builder/vendor/github.com/openshift/api/build/v1.CommonSpec. If set, this field should be validated as having exactly either two or three non-empty fields separated by "/" characters, containing either a combination of OS/Architecture or OS/Architecture/Variant.
+* Add a target-platforms-which-are-allowed-to-fail []string field to github.com/openshift/builder/vendor/github.com/openshift/api/build/v1.CommonSpec. The contents of this field should always be a subset of the contents of the target platforms list, if the target platforms list is set.
```

Comment:
I think that this is an interesting space to explore but I don't have a concept for its priority. I know that ACM has (or at least had) some containers which could not be built for specific architectures. There may be some functionality that cannot be currently supported on an architecture.

Another interesting related possibility here is the ability to specify a separate Dockerfile/build configuration for a subset of architectures requested. This would simplify maintenance for some containers as well. On more than one occasion, I have seen an entirely different forked pathway for s390x. Maintibility would (potentially) be drastically improved if a separate file were leveraged.

Reply:
> Another interesting related possibility here is the ability to specify a separate Dockerfile/build configuration for a subset of architectures requested. This would simplify maintenance for some containers as well. On more than one occasion, I have seen an entirely different forked pathway for s390x. Maintibility would (potentially) be drastically improved if a separate file were leveraged.

like @bparees mentioned, since this is primarily for the internal use case of CI, this would probably be something that we can explore subsequently and not add to this proposal, but yes i see potential value for having separate dockerfiles for arches.

Reply:
> i see potential value for having separate dockerfiles for arches.

I see the value, but i also see it as another order of magnitude of complexity to add.  Multiarch is adding one order of magnitude that i'm already skeptical about (vs investing in other build technologies to get this capability). 

The end state of this road is multiple independent buildconfigs (so you can specify an entirely separate repo, or branch, or commit, and possibly other unique configurations that vary by architecture, and not just a different dockerfile) that are run together to push to a single manifestlist.  And if we think we're going to build that, i'd *really* ask why we're not doing this via tekton/shipwright instead.

Reply:
That's fair. Something for _me_ to take into account instead of vanilla OCP. :)